### **Student Information**
Name: Eliška Kopecká (Eliska Kopecka)

Student ID: 91499149X

GitHub ID: username: elakop; numerical ID: 183389250

Kaggle name: elakop (Eliška Kopecká) <- I can see both there, I'm not sure which one you require; with email elis.kop.b@gmail.com

Kaggle private scoreboard snapshot: I took the picture  1.12.2025 at 23:22 (the end of the competition)

![ranking](Screenshot_2025-12-01_232123.png)

Today, the score changed. There were some late submissions (screenshot from 3.12.2025):

![ranking2](Screenshot2025-12-03220153.png)

---

# **Instructions**

For this lab we have divided the assignments into **three phases/parts**. The `first two phases` refer to the `exercises inside the Master notebooks` of the [DM2025-Lab2-Exercise Repo](https://github.com/difersalest/DM2025-Lab2-Exercise.git). The `third phase` refers to an `internal Kaggle competition` that we are gonna run among all the Data Mining students. Together they add up to `100 points` of your grade. There are also some `bonus points` to be gained if you complete `extra exercises` in the lab **(bonus 15 pts)** and in the `Kaggle Competition report` **(bonus 5 pts)**.

**Environment recommendations to solve lab 2:**
- **Phase 1 exercises:** Need GPU for training the models explained in that part, if you don't have a GPU in your laptop it is recommended to run in Colab or Kaggle for a faster experience, although with CPU they can still be solved but with a slower execution.
- **Phase 2 exercises:** We use Gemini's API so everything can be run with only CPU without a problem.
- **Phase 3 exercises:** For the competition you will probably need GPU to train your models, so it is recommended to use Colab or Kaggle if you don't have a laptop with a dedicated GPU.
- **Optional Ollama Notebook (not graded):** You need GPU, at least 4GB of VRAM with 16 GB of RAM to run the local open-source LLM models. 

## **Phase 1 (30 pts):**

1. __Main Exercises (25 pts):__ Do the **take home exercises** from Sections: `1. Data Preparation` to `9. High-dimension Visualization: t-SNE and UMAP`, in the [DM2025-Lab2-Master-Phase_1 Notebook](https://github.com/difersalest/DM2025-Lab2-Exercise/blob/main/DM2025-Lab2-Master-Phase_1.ipynb). Total: `8 exercises`. Commit your code and submit the repository link to NTU Cool **`BEFORE the deadline (Nov. 3th, 11:59 pm, Monday)`**

2. **Code Comments (5 pts):** **Tidy up the code in your notebook**. 

## **Phase 2 (30 pts):**

1. **Main Exercises (25 pts):** Do the remaining **take home exercises** from Section: `2. Large Language Models (LLMs)` in the [DM2025-Lab2-Master-Phase_2_Main Notebook](https://github.com/difersalest/DM2025-Lab2-Exercise/blob/main/DM2025-Lab2-Master-Phase_2_Main.ipynb). Total: `5 exercises required from sections 2.1, 2.2, 2.4 and 2.6`. Commit your code and submit the repository link to NTU Cool **`BEFORE the deadline (Nov. 24th, 11:59 pm, Monday)`**

2. **Code Comments (5 pts):** **Tidy up the code in your notebook**. 

3. **`Bonus (15 pts):`** Complete the bonus exercises in the [DM2025-Lab2-Master-Phase_2_Bonus Notebook](https://github.com/difersalest/DM2025-Lab2-Exercise/blob/main/DM2025-Lab2-Master-Phase_2_Bonus.ipynb) and [DM2025-Lab2-Master-Phase_2_Main Notebook](https://github.com/difersalest/DM2025-Lab2-Exercise/blob/main/DM2025-Lab2-Master-Phase_2_Main.ipynb) `where 2 exercises are counted as bonus from sections 2.3 and 2.5 in the main notebook`. Total: `7 exercises`. Commit your code and submit the repository link to NTU Cool **`BEFORE the deadline (Nov. 24th, 11:59 pm, Monday)`**

## **Phase 3 (40 pts):**

1. **Kaggle Competition Participation (30 pts):** Participate in the in-class **Kaggle Competition** regarding Emotion Recognition on Twitter by clicking in this link: **[Data Mining Class Kaggle Competition](https://www.kaggle.com/t/3a2df4c6d6b4417e8bf718ed648d7554)**. The scoring will be given according to your place in the Private Leaderboard ranking: 
    - **Bottom 40%**: Get 20 pts of the 30 pts in this competition participation part.

    - **Top 41% - 100%**: Get (0.6N + 1 - x) / (0.6N) * 10 + 20 points, where N is the total number of participants, and x is your rank. (ie. If there are 100 participants and you rank 3rd your score will be (0.6 * 100 + 1 - 3) / (0.6 * 100) * 10 + 20 = 29.67% out of 30%.)   
    Submit your last submission **`BEFORE the deadline (Nov. 24th, 11:59 pm, Monday)`**. Make sure to take a screenshot of your position at the end of the competition and store it as `pic_ranking.png` under the `pics` folder of this repository and rerun the cell **Student Information**.

2. **Competition Report (10 pts)** A report section to be filled in inside this notebook in Markdown Format, we already provided you with the template below. You need to describe your work developing the model for the competition. The report should include a section describing briefly the following elements: 
* Your preprocessing steps.
* The feature engineering steps.
* Explanation of your model.

* **`Bonus (5 pts):`**
    * You will have to describe more detail in the previous steps.
    * Mention different things you tried.
    * Mention insights you gained. 

[Markdown Guide - Basic Syntax](https://www.markdownguide.org/basic-syntax/)

**`Things to note for Phase 3:`**

* **The code used for the competition should be in this Jupyter Notebook File** `DM2025-Lab2-Homework.ipynb`.

* **Push the code used for the competition to your repository**.

* **The code should have a clear separation for the same sections of the report, preprocessing, feature engineering and model explanation. Briefly comment your code for easier understanding, we provide a template at the end of this notebook.**

* Showing the kaggle screenshot of the ranking plus the code in this notebook will ensure the validity of your participation and the report to obtain the corresponding points.

After the competition ends you will have two days more to submit the `DM2025-Lab2-Homework.ipynb` with your report in markdown format and your code. Do everything **`BEFORE the deadline (Nov. 26th, 11:59 pm, Wednesday) to obtain 100% of the available points.`**

Upload your files to your repository then submit the link to it on the corresponding NTU Cool assignment.

## **Deadlines:**

![lab2_deadlines](./pics/lab2_deadlines.png)

---

Next you will find the template report with some simple markdown syntax explanations, use it to structure your content.

You can delete the syntax suggestions after you use them.

---

***

# **Project Report**

**Syntax:** `#` creates the largest heading (H1).

---
**Syntax:** `---` creates a horizontal rule (a separator line).

## 1. Model Development (10 pts Required)

**Syntax:** `##` creates a secondary heading (H2).

**Describe briefly each section, you can add graphs/charts to support your explanations.**

### 1.1 Preprocessing Steps

**Syntax:** `###` creates a tertiary heading (H3).

[Content for Preprocessing]

**Example Syntax for Content:**
*   **Bold text:** `**text**`
*   *Italic text*: `*text*`
*   Bullet point list:
    * Item 1
    * Item 2

Markdown Syntax to Add Image: `![Description of the Image](./your_local_folder/name_of_the_image.png)`

![Example Markdown Syntax to Add Image](./pics/example_md_img.png)

### 1.2 Feature Engineering Steps

[Content for Feature Engineering]

### 1.3 Explanation of Your Model

[Content for Model Explanation]

---

## 2. Bonus Section (5 pts Optional)

**Add more detail in previous sections**

### 2.1 Mention Different Things You Tried

[Content for Experiments]

### 2.2 Mention Insights You Gained

[Content for Insights]

---

**`From here on starts the code section for the competition.`**

---

# **Competition Code**

## 1. Preprocessing Steps

---
### Required Libraries

In [1]:
import json
import pandas as pd
import numpy as np
import re

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from scipy.sparse import hstack


### Files and paths

**Note:** I was orriginally runnig the code without jupiter notebook and these are the files I used

I'm not sure if you will run my code, if yes I'll try to post the source files to github LAB2-EXCERCISE,
but you'll need to change the DATA_DIR variable.

In [3]:
# This is the file that I use for this lab
# Maybe you will need to chnage the source of the data
DATA_DIR = "C:/Users/Administrator/Documents/MUNI/2/NTNU_fall/Data Mining/DM2025Labs/DM2025-Lab2-Exercise"

JSON_FILE = f"{DATA_DIR}/final_posts.json"
EMOTION_FILE = f"{DATA_DIR}/emotion.csv"
SPLIT_FILE = f"{DATA_DIR}/data_identification.csv"
OUTPUT_FILE = f"{DATA_DIR}/submission.csv"

---
Preprocessing steps

This section:
- loads the data files
- Merge them into a single DataFrame
- Clean and preprocess the text data

There are three data sources `final_posts.json`, which contains the actual text of posts (JSON structure), `emotion.csv`, which contains emotion labels for training data, and `data_identification.csv` with identifies which posts are for training vs testing

With text files, I usually use encoding=utf-8 <- it actually happend to me, that I needed to add it to one of the Labs before.
The JSON file has nested structure, actual data is buried inside `root` → `_source` → `post`. I had to look at the file first to figure out how to get to the `post_id` and `text` fields. That's why I'm using `item['root']['_source']['post']['post_id']` instead of just `item['post_id']`.

In [4]:
# Load JSON file with posts
with open(JSON_FILE, 'r', encoding='utf-8') as f:
    posts_data = json.load(f)

# Extract post_id and text from nested JSON structure
posts = []
for item in posts_data:
    post_id = item['root']['_source']['post']['post_id']
    text = item['root']['_source']['post']['text']
    posts.append({'id': post_id, 'text': text})

posts_df = pd.DataFrame(posts)
print(f"{len(posts_df)} posts from JSON")
posts_df.head()

64171 posts from JSON


id                                               text
0  0x61fc95  We got the ranch, loaded our guns and sat up t...
1  0x35663e  I bet there is an army of married couples who ...
2  0xc78afe                         This could only end badly.
3  0x90089c  My sister squeezed a lime in her milk when she...
4  0xaba820         and that got my head bobbing a little bit.

Loading the CSV with emotion labels (anger, joy, fear, etc.) for each post.

In [5]:
# Load emotion labels
emotion_df = pd.read_csv(EMOTION_FILE)
print(f"{len(emotion_df)} emotion labels")
emotion_df.head()

47890 emotion labels


id emotion
0  0x35663e     joy
1  0xc78afe    fear
2  0x90089c     joy
3  0x2ffb63     joy
4  0x989146     joy

This file tells us which posts are for training and which are for testing. It's needed because the main JSON file contains all posts mixed together.

In [6]:
# Load train/test split information
split_df = pd.read_csv(SPLIT_FILE)
print(f"Split information for {len(split_df)} posts")
split_df['split'].value_counts()

Split information for 64171 posts


split
train    47890
test     16281
Name: count, dtype: int64

Below I'm merging all three dataframes together using the `id` column. After merging, I have everything in one place.The text, the train/test split info, and the emotion labels are all under `df` variable. Test posts will have `NaN` in the emotion column since I don't know their labels.

In [7]:
# Merge all data together
df = posts_df.merge(split_df, on='id', how='left')
df = df.merge(emotion_df, on='id', how='left')

print(f"Total posts: {len(df)}")
print(f"Train posts: {len(df[df['split'] == 'train'])}")
print(f"Test posts: {len(df[df['split'] == 'test'])}")
df.head()

Total posts: 64171
Train posts: 47890
Test posts: 16281


id                                               text  split emotion
0  0x61fc95  We got the ranch, loaded our guns and sat up t...   test     NaN
1  0x35663e  I bet there is an army of married couples who ...  train     joy
2  0xc78afe                         This could only end badly.  train    fear
3  0x90089c  My sister squeezed a lime in her milk when she...  train     joy
4  0xaba820         and that got my head bobbing a little bit.   test     NaN

Checking how emotions are distributed in the training data. This is important because the classes are very imbalanced. Joy is about 50% of the data while disgust is only 2%. This might affect which model that is going to be chosen or how it will be configured.

In [8]:
# Check emotion distribution in training data
df[df['split'] == 'train']['emotion'].value_counts()

emotion
joy         23797
anger       10694
surprise     6281
sadness      3926
fear         2009
disgust      1183
Name: count, dtype: int64

Text preprocessing

For NLP models to work well, I need to clean the text first.
I'm using `re` library (regular expressions) to remove URLs, handle @mentions and #hashtags, etc.
I noticed that things like `!!!` or `???` or TYPING IN CAPS might actually be useful for emotion detection, so I'm converting them to text markers instead of just deleting them.

Steps applyed:
- Convert to lowercase
- Remove URLs
- Handle mentions (@user) and hashtags (#topic)
- Convert punctuation patterns to text markers (preserve emotion signals)
- Remove extra whitespace

The function below cleans the text step by step using regex (`re.sub`).
I first save the original to detect CAPS (before lowercasing), then remove URLs, convert @mentions and #hashtags to markers, and handle punctuation.

The markers at the end ("veryexcited", "shouting") preserve the emotional signals from the original text.

"emptytext" is a fallback for emoji-only posts that would otherwise become empty strings.
<- wich actually caused error with my first submission

In [10]:
def preprocess_text(text):

    if pd.isna(text):
        return ""
    
    original = text
    text = text.lower()
    
    # Track emotion signals before removing them
    has_exclamations = text.count('!') >= 2
    has_questions = text.count('?') >= 2
    has_caps = sum(1 for c in original if c.isupper()) > len(original) * 0.3
    
    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)
    
    # convert mentions and hashtags to markers
    text = re.sub(r'@\w+', ' mention ', text)
    text = re.sub(r'#(\w+)', r' hashtag \1 ', text)
    
    # Convert punctuation patterns to text markers
    text = re.sub(r'!+', ' exclamation ', text)
    text = re.sub(r'\?+', ' question ', text)
    
    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text).strip()
    
    # Add emotion markers at the end
    if has_exclamations:
        text = text + " veryexcited"
    if has_questions:
        text = text + " questioning"
    if has_caps:
        text = text + " shouting"
    
    # Handle empty texts (e.g., emoji-only posts)
    return text if text.strip() else "emptytext"

# Test the preprocessing function
test_texts = [
    "I HATE THIS!!! Why won't it work??",
    "Check out @user and #happiness",
    "Visit https://example.com for more info",
    "😘 ☂️☂️"  # Emoji-only post
]

#Preprocessing Examples
print("PREPROCESS. EXAMPLES")
for t in test_texts:
    print(f"Original: {t}")
    print(f"Cleaned:  {preprocess_text(t)}")
    print()

PREPROCESS. EXAMPLES
Original: I HATE THIS!!! Why won't it work??
Cleaned:  i hate this exclamation why won't it work question veryexcited questioning

Original: Check out @user and #happiness
Cleaned:  check out mention and hashtag happiness

Original: Visit https://example.com for more info
Cleaned:  visit for more info

Original: 😘 ☂️☂️
Cleaned:  😘 ☂️☂️



With the code bellow I apply the preprocessing function to all 64k posts. The `.apply()` method runs the function on each row.
I'm also checking how many posts ended up as "emptytext" (the emoji-only ones).

In [11]:
# Apply preprocessing to all texts
df['clean_text'] = df['text'].apply(preprocess_text)

# Check for empty texts
empty_count = (df['clean_text'] == 'emptytext').sum()
print(f"Empty texts after preprocessing: {empty_count}")

# Show some examples
print("\n SAMPLE: Preprocessed Texts")
df[['text', 'clean_text']].head()

Empty texts after preprocessing: 0

 SAMPLE: Preprocessed Texts


text  \
0  We got the ranch, loaded our guns and sat up t...   
1  I bet there is an army of married couples who ...   
2                         This could only end badly.   
3  My sister squeezed a lime in her milk when she...   
4         and that got my head bobbing a little bit.   

                                          clean_text  
0  we got the ranch, loaded our guns and sat up t...  
1  i bet there is an army of married couples who ...  
2                         this could only end badly.  
3  my sister squeezed a lime in her milk when she...  
4         and that got my head bobbing a little bit.

## 2. Feature Engineering Steps

This section focuses on creating features for classification model:
- **TF-IDF features** - word-level n-grams (1-3 grams)
- **Character n-grams** - capture writing style patterns (3-6 char n-grams)
- **Manual features** - text statistics and emotion keyword counts

Manual Features

- Text length and word count
- Punctuation counts (exclamation, question marks)
- Capital letter ratio
- Emoji count
- Emotion keyword matching (simple lexicon-based approach)

Besides TF-IDF, I wanted to add some hand-crafted features that might help with emotion detection. Things like exclamation marks, CAPS, or words like "hate" and "love" are obvious emotion signals.

The `+1` in `capital_ratio` is to avoid division by zero for empty texts.

For emoji detection, I'm counting any character with unicode value above 127. It's not perfect (catches accented letters too), but good enough.

In [12]:
def extract_features(row):

    text = row['text'] if pd.notna(row['text']) else ""
    features = {}
    
    # Basic text statistics
    features['length'] = len(text)
    features['word_count'] = len(text.split())
    
    # Punctuation features (emotion indicators)
    features['exclamation'] = text.count('!')
    features['question'] = text.count('?')
    
    # Capitalization (anger/excitement indicator)
    features['capital_ratio'] = sum(1 for c in text if c.isupper()) / (len(text) + 1)
    
    # Emoji count (unicode characters above ASCII)
    features['emoji_count'] = len([c for c in text if ord(c) > 127])
    
    # Emotion keyword matching (simple lexicon approach)
    text_lower = text.lower()
    
    # Anger keywords
    anger_words = ['hate', 'angry', 'mad', 'stupid', 'damn', 'hell', 'furious']
    features['anger_kw'] = sum(1 for w in anger_words if w in text_lower)
    
    # Joy keywords
    joy_words = ['happy', 'love', 'great', 'haha', 'lol', 'amazing', 'wonderful']
    features['joy_kw'] = sum(1 for w in joy_words if w in text_lower)
    
    # Sadness keywords
    sad_words = ['sad', 'sorry', 'cry', 'miss', 'depressed', 'alone']
    features['sad_kw'] = sum(1 for w in sad_words if w in text_lower)
    
    # Fear keywords
    fear_words = ['scared', 'afraid', 'fear', 'worried', 'panic', 'nervous']
    features['fear_kw'] = sum(1 for w in fear_words if w in text_lower)
    
    # Surprise keywords
    surprise_words = ['wow', 'omg', 'what', 'really', 'wtf', 'shocked']
    features['surprise_kw'] = sum(1 for w in surprise_words if w in text_lower)
    
    return features

# Extract features for all rows
feature_data = df.apply(extract_features, axis=1)
feature_df = pd.DataFrame(feature_data.tolist())

# Combine with main dataframe
df = pd.concat([df, feature_df], axis=1)

print(f"Extracted {len(feature_df.columns)} manual features")
print(f"Feature names: {list(feature_df.columns)}")
feature_df.describe()

Extracted 11 manual features
Feature names: ['length', 'word_count', 'exclamation', 'question', 'capital_ratio', 'emoji_count', 'anger_kw', 'joy_kw', 'sad_kw', 'fear_kw', 'surprise_kw']


length    word_count   exclamation      question  capital_ratio  \
count  64171.000000  64171.000000  64171.000000  64171.000000   64171.000000   
mean      76.742688     14.104860      0.212261      0.134640       0.051056   
std       43.342436      7.657896      0.658740      0.435182       0.077572   
min        2.000000      1.000000      0.000000      0.000000       0.000000   
25%       43.000000      8.000000      0.000000      0.000000       0.019231   
50%       72.000000     14.000000      0.000000      0.000000       0.032787   
75%      105.000000     19.000000      0.000000      0.000000       0.055556   
max      703.000000     89.000000     13.000000     12.000000       0.918033   

        emoji_count      anger_kw        joy_kw        sad_kw       fear_kw  \
count  64171.000000  64171.000000  64171.000000  64171.000000  64171.000000   
mean       0.343816      0.058531      0.129560      0.047327      0.015989   
std        1.072932      0.241939      0.356135      0.217773      0.128258   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max       73.000000      3.000000      3.000000      4.000000      2.000000   

        surprise_kw  
count  64171.000000  
mean       0.114974  
std        0.332578  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        3.000000

The code bellow is splitting the data based on the `split` column that has been loaded earlier. Using `.copy()` to avoid the "SettingWithCopyWarning" from pandas.

In [13]:
# Split into train and test based on the split column
train_df = df[df['split'] == 'train'].copy()
test_df = df[df['split'] == 'test'].copy()

print(f"Training samples: {len(train_df)}")
print(f"Test samples: {len(test_df)}")

Training samples: 47890
Test samples: 16281


TF-IDF Vectorization

Creating of two types of TF-IDF features:
1. **Word-level TF-IDF** with n-grams (1, 2, 3) - captures word meanings and short phrases
2. **Character-level TF-IDF** with n-grams (3-6) - captures writing style and morphological patterns

[sklearn TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html)

TF-IDF converts text into numbers. Each word/phrase gets a score based on how important it is in a document vs. the whole dataset.
I'm using `ngram_range=(1,3)` to capture single words but also phrases like "so happy" or "hate this so". The `max_features=10000` limits vocabulary size so it doesn't get too big. `max_df=0.85` ignores words that appear in more than 85% of documents (like "the", "is") since they're not useful for classification.
The `fit_transform` on training data learns the vocabulary, then I just `transform` the test data using the same vocabulary.

In [14]:
# Word-level TF-IDF vectorizer
tfidf_word = TfidfVectorizer(
    max_features=10000, # maximum vocabulary size
    ngram_range=(1, 3), # Unigrams, bigrams, trigrams
    min_df=2,   # minimum document frequency
    max_df=0.85,    # Maximum document frequency (ignore very common words)
    sublinear_tf=True   # sublinear tf scaling (1 + log(tf))
)

# Fit on training data, transform both train and test
X_train_word = tfidf_word.fit_transform(train_df['clean_text'])
X_test_word = tfidf_word.transform(test_df['clean_text'])

print(f"Word TF-IDF shape: {X_train_word.shape}")

Word TF-IDF shape: (47890, 10000)


Same idea as before but at character level. Vectorizer looks at letter patterns instead of words. This can catch typos, repeated letters ("nooo"), and writing style. The `analyzer='char'` switches from word to character mode.

In [15]:
# Character-level TF-IDF vectorizer
# This part should take approx. 40s
tfidf_char = TfidfVectorizer(
    max_features=5000,  # maximum vocabulary size
    ngram_range=(3, 6), # Character n-grams from 3 to 6
    analyzer='char',    # character-level analysis
    sublinear_tf=True
)

# Fit on training data, transform both train and test
X_train_char = tfidf_char.fit_transform(train_df['clean_text'])
X_test_char = tfidf_char.transform(test_df['clean_text'])

print(f"Character TF-IDF shape: {X_train_char.shape}")

Character TF-IDF shape: (47890, 5000)


Combine all features

Getting the manual features into arrays. StandardScaler normalizes them so they're on the same scale as TF-IDF values, otherwise features like `length` (values in hundreds) would dominate over smaller features.

In [16]:
# Get manual features and scale them
feature_cols = list(feature_df.columns)
X_train_manual = train_df[feature_cols].values
X_test_manual = test_df[feature_cols].values

# Scale manual features
scaler = StandardScaler()
X_train_manual = scaler.fit_transform(X_train_manual)
X_test_manual = scaler.transform(X_test_manual)

print(f"Manual features shape: {X_train_manual.shape}")

Manual features shape: (47890, 11)


Combining all three feature sets (word TF-IDF, character TF-IDF, and manual features) into one matrix.

In [17]:
# Combine all features using scipy hstack (for sparse matrices)
X_train = hstack([X_train_word, X_train_char, X_train_manual])
X_test = hstack([X_test_word, X_test_char, X_test_manual])

# Get labels
y_train = train_df['emotion'].values

print(f"Training features shape: {X_train.shape}")
print(f"Test features shape: {X_test.shape}")
print(f"Total features: {X_train.shape[1]}")

Training features shape: (47890, 15011)
Test features shape: (16281, 15011)
Total features: 15011


## 3. Model Implementation Steps

This section focuses on:
- Training multiple classification models
- Evaluating using cross-validation (F1 Macro score)
- Selecting the best model
- Generating predictions for the test set

First I try Logistic Regression model

The `class_weight='balanced'` is here because teh data is imbalanced (50% joy, only 2% disgust). This tells the model to pay more attention to rare classes.
`C=2.0` controls regularization. Higher value means less regularization. `n_jobs=-1` uses all CPU cores to speed things up.
I'm using 3-fold cross-validation with F1 Macro score because that's how the competition evaluates submissions. This gives a more reliable estimate than just training once.

In [18]:
# Model 1: Logistic Regression
# This part takes approx. 2m 30s
lr_model = LogisticRegression(
    C=2.0,  # inverse of regularization strength
    max_iter=1000,  # maximum iterations
    class_weight='balanced',    #for handling class imbalance
    random_state=42,
    n_jobs=-1
)

lr_model.fit(X_train, y_train)

# Cross-validation with F1 Macro (competition metric)
lr_cv_scores = cross_val_score(lr_model, X_train, y_train, cv=3, scoring='f1_macro', n_jobs=-1)
print("Training Model 1: Logistic Regression")
print(f"CV F1 Scores: {lr_cv_scores}")
print(f"Mean CV F1 Score: {lr_cv_scores.mean():.4f} (+/- {lr_cv_scores.std():.4f})")

Training Model 1: Logistic Regression
CV F1 Scores: [0.44767661 0.44679706 0.44104679]
Mean CV F1 Score: 0.4452 (+/- 0.0029)


Second model is Linear SVM

Same idea with `class_weight='balanced'`. The `C` parameter is a bit lower (0.8) which means slightly more regularization. I increased `max_iter` to 2000 because SVM may need more iterations to converge.

In [19]:
# Model 2: Linear SVM
# This part takes about 3 m 10 s
svm_model = LinearSVC(
    C=0.8, # Regularization parameter
    class_weight='balanced',
    max_iter=2000,
    random_state=42
)

svm_model.fit(X_train, y_train)

svm_cv_scores = cross_val_score(svm_model, X_train, y_train, cv=3, scoring='f1_macro', n_jobs=-1)
print("Model 2: Linear SVM")
print(f"CV F1 Scores: {svm_cv_scores}")
print(f"Mean CV F1 Score: {svm_cv_scores.mean():.4f} (+/- {svm_cv_scores.std():.4f})")

Model 2: Linear SVM
CV F1 Scores: [0.43805779 0.43739666 0.43555725]
Mean CV F1 Score: 0.4370 (+/- 0.0011)


Third model is Random Forest. An ensemble of 300 decision trees.

`max_depth=25` limits how deep each tree can grow. Without this, trees could overfit to the training data.

In [20]:
# Model 3: Random Forest
# This part took 2 m 15 s
rf_model = RandomForestClassifier(
    n_estimators=300,   # Number of trees
    max_depth=25,   # maximum tree depth
    class_weight='balanced',
    random_state=42,
    n_jobs=-1
)

rf_model.fit(X_train, y_train)

rf_cv_scores = cross_val_score(rf_model, X_train, y_train, cv=3, scoring='f1_macro', n_jobs=-1)
print("Model 3: Random Forest")
print(f"CV F1 Scores: {rf_cv_scores}")
print(f"Mean CV F1 Score: {rf_cv_scores.mean():.4f} (+/- {rf_cv_scores.std():.4f})")

Model 3: Random Forest
CV F1 Scores: [0.37630619 0.39009887 0.38924469]
Mean CV F1 Score: 0.3852 (+/- 0.0063)


Selecting the best model

Comparing all three models by their CV F1 score and picking the best one. The winner will be used for final predictions.

In [21]:
# Compare all models
models = {
    'Logistic Regression': (lr_model, lr_cv_scores.mean()),
    'Linear SVM': (svm_model, svm_cv_scores.mean()),
    'Random Forest': (rf_model, rf_cv_scores.mean())
}

# Print comparison
for name, (model, score) in models.items():
    print(f"{name:25s}: F1 = {score:.4f}")

# Select best model
best_name = max(models, key=lambda x: models[x][1])
best_model, best_score = models[best_name]

print("\n")
print(f" BEST MODEL: {best_name}")
print(f" Cross-validation F1 Score: {best_score:.4f}")

Logistic Regression      : F1 = 0.4452
Linear SVM               : F1 = 0.4370
Random Forest            : F1 = 0.3852


 BEST MODEL: Logistic Regression
 Cross-validation F1 Score: 0.4452


Generating predction

Running the best model on test data. Checking the prediction distribution to make sure it looks reasonable (not predicting just one class).

In [22]:
# Generate predictions using the best model
test_predictions = best_model.predict(X_test)

print(f"Generated {len(test_predictions)} predictions")

# Check prediction distribution
print("\nPrediction Distribution:")
pred_dist = pd.Series(test_predictions).value_counts()
print(pred_dist)

Generated 16281 predictions

Prediction Distribution:
joy         5772
anger       3730
surprise    2478
sadness     1906
fear        1606
disgust      789
Name: count, dtype: int64


Creating submission file

Creating the submission file in the format required by the competition. Which is two columns: `id` and `emotion`.

In [23]:
# Create submission DataFrame
submission_df = pd.DataFrame({
    'id': test_df['id'],
    'emotion': test_predictions
})

# Verify submission format
print("Submission Preview:")
print(submission_df.head(10))
print(f"\nTotal rows: {len(submission_df)}")

Submission Preview:
          id   emotion
0   0x61fc95      fear
4   0xaba820      fear
5   0x66e44d       joy
6   0xc03cf5       joy
8   0x02f65a     anger
16  0x479407       joy
20  0xe07a21      fear
27  0x06d186       joy
29  0xa9a658  surprise
41  0x0a0102       joy

Total rows: 16281


Saving to CSV before uploading

In [26]:
# Save submission to CSV
submission_df.to_csv(OUTPUT_FILE, index=False)

print(f"\nFile saved to: {OUTPUT_FILE}")
print(f"Total predictions: {len(submission_df)}")
print(f"\nBest model: {best_name}")


File saved to: C:/Users/Administrator/Documents/MUNI/2/NTNU_fall/Data Mining/DM2025Labs/DM2025-Lab2-Exercise/submission.csv
Total predictions: 16281

Best model: Logistic Regression


---
## Summary

1. **Preprocessing:** Loaded JSON posts, merged with emotion labels and train/test split. Cleaned text by removing URLs, converting @mentions and #hashtags to markers. Preserved emotion signals (!!!, ???, CAPS) as text markers instead of removing them.

2. **Feature Engineering:** Created TF-IDF features with word n-grams (1-3) and character n-grams (3-6). Extracted manual features like text length, punctuation counts, and emotion keyword matching. Combined everything into ~15,000 features.

3. **Model:** Trained Logistic Regression, Linear SVM, and Random Forest. Evaluated with 3-fold cross-validation using F1 Macro. Logistic Regression performed best.

### Explanation of the Model

I compared three models using 3-fold cross-validation with F1 Macro (the competition metric):

| Model | CV F1 Score |
|-------|-------------|
| Logistic Regression | 0.4452 |
| Linear SVM | 0.4370 |
| Random Forest | 0.3852 |

Logistic Regression with `class_weight='balanced'` worked best.

### Different Things I Tried

My first submission used basic TF-IDF with simple preprocessing (just removing URLs and punctuation). 

For the second submission, I made a few improvements:
- Increased TF-IDF vocabulary size and added trigrams
- Added emotion keyword matching (simple lists of words like "hate", "love", "scared")
- Preserved emotion signals (!!!, ???, CAPS) as text markers instead of removing them

### Possible Improvements

- Pretrained embeddings (Word2Vec, GloVe) instead of TF-IDF
- Use deep learning models (BERT)